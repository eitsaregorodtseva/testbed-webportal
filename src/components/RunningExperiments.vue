<template>
<div class="mt-3 mb-4">
  <table class="table table-striped table-sm mt-3" v-if="expList.length">
    <thead>
      <tr>
        <th>Id</th>
        <th>User</th>
        <th>Elapsed</th>
        <th>State</th>
        <th>Ending</th>
        <th>Nodes</th>
      </tr>
    </thead>
    <tbody>
      <tr v-for="exp in expList">
        <td class="nowrap" v-if="isAdmin">
          <router-link :to="{name: 'experimentDetails', params: { id: exp.id }}">{{exp.id}}</router-link>
        </td>
        <td class="nowrap" v-else>{{exp.id}}</td>
        <td class="nowrap" v-if="isAdmin">
          <router-link :to="{name: 'users', query: { search: exp.user }}">{{exp.user}}</router-link>
        </td>
        <td class="nowrap" v-else>{{exp.user}}</td>
        <td class="durationProgress nowrap" :style="`text-align: center; --progress: ${experimentProgress(exp)}%`" :title="showRemaining(exp)">
          {{exp.effective_duration | humanizeDuration}}
          <small class="text-dark">({{experimentProgress(exp)}}%)</small>
        </td>
        <td class="nowrap"><span class="badge badge-state" :class="exp.state | stateBadgeClass">{{exp.state}}</span></td>
        <td class="nowrap" :title="showStarted(exp)">{{exp.start_date | formatDateTime(exp.submitted_duration)}}</td>
        <td style="width: 600px">
          <div class="ellipsis ellipsis-indicator cursor" @click="toggleEllipsis">
            <span class="text-muted nowrap mr-1" v-if="exp.nodes.length <= 1">({{exp.nodes.length}} node)</span>
            <span class="text-muted nowrap mr-1" v-else>({{exp.nodes.length}} nodes)</span>
            <span class="nodes mr-1" :class="{'comma': index + 1 < exp.nodes.length}"
              v-for="(node, index) in [...exp.nodes].sort((a, b) => nodeSortByHostname(a, b))">{{node | stripDomain}}</span>
            <!-- {{exp.nodes.join(', ')}} -->
          </div>
        </td>
      </tr>
    </tbody>
  </table>
  <p class="text-muted font-italic" v-else>
    No experiment currently running
  </p>

</div> <!-- container -->

</template>

<script>
import { auth } from '@/auth'

function getEventPath (event) {
  if (event.composedPath && event.composedPath()) return event.composedPath()
  if (event.path) return event.path

  let path = []
  let target = event.target

  while (target.parentNode) {
    path.push(target)
    target = target.parentNode
  }
  path.push(document, window)
  return path
}

export default {
  name: 'RunningExperiments',

  props: {
    expList: {
      type: Array,
      default: () => [],
    },
  },

  data () {
    return {
      isAdmin: auth.isAdmin,
    }
  },

  methods: {

    showStarted (exp) {
      return `started ${this.$options.filters.formatDateTime(exp.start_date)}`
    },

    showRemaining (exp) {
      return `remaining ${this.$options.filters.humanizeDuration(exp.submitted_duration - exp.effective_duration)}`
    },

    experimentProgress (exp) {
      return Math.min(100, Math.round(100 * exp.effective_duration / exp.submitted_duration))
    },

    nodeSortByHostname (a, b) {
      // Allow sorting of hostnames by domain then by servername
      // the trick is to append the servername after the domain (e.g 'a8-1.grenoble.iot-lab.info' -> 'grenoble.iot-lab.info.a8-1')
      // then it's possible to simply sort the strings
      let sortByHostname = function (node) {
        let chuncks = node.split('.')
        chuncks.push(chuncks.shift()) // or we could use chunks.reverse()
        return chuncks.join('.')
      }
      return sortByHostname(a).localeCompare(sortByHostname(b))
    },

    displayNodes (exp) {
      return exp.nodes.map(n => this.$options.filters.stripDomain(n)).sort().map(n => `<i>${n}</i>`).join(', ')
    },

    toggleEllipsis (event) {
      getEventPath(event).find(e => e.tagName === 'DIV').classList.toggle('ellipsis')
    },

  },

}
</script>

<style scoped>
.nowrap {
  white-space: nowrap;
}
td.nowrap {
  vertical-align: middle;
}
.ellipsis {
  max-width: calc(600px - 0.6rem);
  overflow: hidden;
  text-overflow: hidden;
  white-space: nowrap;
}
.durationProgress::after,.durationProgress::before {
  top: calc(50% - 12px);
  height: 24px;
}
.nodes {
  white-space: nowrap;
  display: inline-block;
}
.nodes.comma::after {
  content: ", ";
}
.ellipsis-indicator {
  position: relative;
}
.ellipsis-indicator:hover::after {
  position: absolute;
  right: 0;
  top: calc(50% - 12px);
  /*top: 50%;*/
  top: 0;
  padding: 0 5px;
  font-family: fontawesome;
  content: "\f141";
  content: "\f0d7";
  content: "\f065";
  color: var(--dark);
  font-size: 1em;
  background: #ccc;
  border-radius: 5px;
}
</style>
