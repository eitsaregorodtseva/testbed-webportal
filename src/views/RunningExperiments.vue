<template>
<div class="container mt-3">
  <h2>Running experiments</h2>
  <table class="table table-striped table-sm mt-3">
    <thead>
      <tr>
        <th>Id</th>
        <th>User</th>
        <th>Elapsed</th>
        <th>State</th>
        <th>Ending</th>
        <th>Nodes</th>
      </tr>
    </thead>
    <tbody>
      <tr v-for="exp in runningExperiments">
        <td class="nowrap">{{exp.id}}</td>
        <td class="nowrap">{{exp.user}}</td>
        <td class="durationProgress nowrap" :style="`text-align: center; --progress: ${experimentProgress(exp)}%`" :title="showRemaining(exp)">
          {{exp.effective_duration | humanizeDuration}}
          <small class="text-dark">({{experimentProgress(exp)}}%)</small>
        </td>
        <td class="nowrap"><span class="badge badge-state" :class="exp.state | stateBadgeClass">{{exp.state}}</span></td>
        <td class="nowrap" :title="showStarted(exp)">{{exp.start_date | formatDateTime(exp.submitted_duration)}}</td>
        <td style="width: 600px">
          <div class="ellipsis ellipsis-indicator cursor" @click="toggleEllipsis">
            <span class="text-muted nowrap mr-1" v-if="exp.nodes.length <= 1">({{exp.nodes.length}} node)</span>
            <span class="text-muted nowrap mr-1" v-else>({{exp.nodes.length}} nodes)</span>
            <span class="nodes mr-1" :class="{'comma': index + 1 < exp.nodes.length}"
              v-for="(node, index) in [...exp.nodes].sort((a, b) => nodeSortByHostname(a, b))">{{node | stripDomain}}</span>
            <!-- {{exp.nodes.join(', ')}} -->
          </div>
        </td>
      </tr>
    </tbody>
  </table>
  <p class="text-muted">
    {{runningExperiments.length}} running experiments
  </p>

</div> <!-- container -->

</template>

<script>
import { iotlab } from '@/rest'

export default {
  name: 'RunningExperiments',

  data () {
    return {
      runningExperiments: [],
    }
  },

  created () {
    iotlab.getRunningExperiments().then(data => { this.runningExperiments = data }).catch(err => {
      this.$notify({text: err.response.data.message || 'Failed to fetch running experiments', type: 'error'})
    })
  },

  methods: {

    showStarted (exp) {
      return `started ${this.$options.filters.formatDateTime(exp.start_date)}`
    },

    showRemaining (exp) {
      return `remaining ${this.$options.filters.humanizeDuration(exp.submitted_duration - exp.effective_duration)}`
    },

    experimentProgress (exp) {
      return Math.min(100, Math.round(100 * exp.effective_duration / exp.submitted_duration))
    },

    nodeSortByHostname (a, b) {
      // Allow sorting of hostnames by domain then by servername
      // the trick is to append the servername after the domain (e.g 'a8-1.grenoble.iot-lab.info' -> 'grenoble.iot-lab.info.a8-1')
      // then it's possible to simply sort the strings
      let sortByHostname = function (node) {
        let chuncks = node.split('.')
        chuncks.push(chuncks.shift()) // or we could use chunks.reverse()
        return chuncks.join('.')
      }
      return sortByHostname(a).localeCompare(sortByHostname(b))
    },

    displayNodes (exp) {
      return exp.nodes.map(n => this.$options.filters.stripDomain(n)).sort().map(n => `<i>${n}</i>`).join(', ')
    },

    toggleEllipsis (event) {
      event.path.find(e => e.tagName === 'DIV').classList.toggle('ellipsis')
    },

  },

}
</script>

<style scoped>
.nowrap {
  white-space: nowrap;
}
td.nowrap {
  vertical-align: middle;
}
.ellipsis {
  max-width: calc(600px - 0.6rem);
  overflow: hidden;
  text-overflow: hidden;
  white-space: nowrap;
}
.durationProgress::after,.durationProgress::before {
  top: calc(50% - 12px);
  height: 24px;
}
.nodes {
  white-space: nowrap;
  display: inline-block;
}
.nodes.comma::after {
  content: ", ";
}
.ellipsis-indicator {
  position: relative;
}
.ellipsis-indicator:hover::after {
  position: absolute;
  right: 0;
  top: calc(50% - 12px);
  /*top: 50%;*/ 
  top: 0;
  padding: 0 5px;
  font-family: fontawesome;
  content: "\f141";
  content: "\f0d7";
  content: "\f065";
  color: var(--dark);
  font-size: 1em;
  background: #ccc;
  border-radius: 5px;
}
</style>
