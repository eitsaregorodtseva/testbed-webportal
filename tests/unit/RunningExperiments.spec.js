import { shallowMount, createLocalVue } from '@vue/test-utils'
import VueRouter from 'vue-router'
import RunningExperiments from '@/components/RunningExperiments'
import * as Filters from '@/filters'
import { iotlab } from '@/rest'
jest.mock('@/rest')

const router = new VueRouter()
const localVue = createLocalVue()
localVue.use(VueRouter)
localVue.filter('humanizeDuration', Filters.humanDuration)

describe('RunningExperiments.vue', () => {

  it('should match snapshot', () => {
    const propsData = {
      expList: [{"submitted_duration":20,"submission_date":"2018-09-25T15:02:57Z","nodes":["m3-12.devgrenoble.iot-lab.info","m3-10.devgrenoble.iot-lab.info","m3-13.devgrenoble.iot-lab.info"],"nb_nodes":3,"effective_duration":0,"name":"","stop_date":"1970-01-01T00:00:00Z","id":13308,"state":"Running","scheduled_date":"2018-09-25T15:02:58Z","user":"bformet","start_date":"2018-09-25T15:02:58Z"},{"submitted_duration":60,"submission_date":"2018-09-25T15:03:03Z","nodes":["a8-1.devgrenoble.iot-lab.info","a8-4.devgrenoble.iot-lab.info","a8-17.devgrenoble.iot-lab.info","a8-3.devgrenoble.iot-lab.info","a8-5.devgrenoble.iot-lab.info","a8-7.devgrenoble.iot-lab.info","a8-9.devgrenoble.iot-lab.info","a8-6.devgrenoble.iot-lab.info","a8-2.devgrenoble.iot-lab.info","a8-8.devgrenoble.iot-lab.info"],"nb_nodes":10,"effective_duration":0,"name":"","stop_date":"1970-01-01T00:00:00Z","id":13309,"state":"Running","scheduled_date":"2018-09-25T15:03:04Z","user":"bformet","start_date":"2018-09-25T15:03:04Z"}],
    }
    const wrapper = shallowMount(RunningExperiments, {localVue, router, propsData})
    expect(wrapper.element).toBeDefined()
    expect(wrapper.element).toMatchSnapshot()
  })

})
