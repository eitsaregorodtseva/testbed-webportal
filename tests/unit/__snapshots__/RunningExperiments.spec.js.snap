// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RunningExperiments.vue should match snapshot 1`] = `
<div
  class="mt-3 mb-4"
>
  <table
    class="table table-striped table-sm mt-3"
  >
    <thead>
      <tr>
        <th>
          Id
        </th>
         
        <th>
          User
        </th>
         
        <th>
          Elapsed
        </th>
         
        <th>
          State
        </th>
         
        <th>
          Ending
        </th>
         
        <th>
          Nodes
        </th>
      </tr>
    </thead>
     
    <tbody>
      <tr>
        <td
          class="nowrap"
        >
          13308
        </td>
         
        <td
          class="nowrap"
        >
          bformet
        </td>
         
        <td
          class="durationProgress nowrap"
          style="text-align: center;"
          title="remaining 20 minutes"
        >
          
          
          
          <small
            class="text-dark"
          >
            (0%)
          </small>
        </td>
         
        <td
          class="nowrap"
        >
          <span
            class="badge badge-state badge-success"
          >
            Running
          </span>
        </td>
         
        <td
          class="nowrap"
          title="started 2018-09-25 17:02"
        >
          2018-09-25 17:22
        </td>
         
        <td
          style="width: 600px;"
        >
          <div
            class="ellipsis ellipsis-indicator cursor"
          >
            <span
              class="text-muted nowrap mr-1"
            >
              (3 nodes)
            </span>
             
            <span
              class="nodes mr-1 comma"
            >
              m3-10.devgrenoble
            </span>
            <span
              class="nodes mr-1 comma"
            >
              m3-12.devgrenoble
            </span>
            <span
              class="nodes mr-1"
            >
              m3-13.devgrenoble
            </span>
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="nowrap"
        >
          13309
        </td>
         
        <td
          class="nowrap"
        >
          bformet
        </td>
         
        <td
          class="durationProgress nowrap"
          style="text-align: center;"
          title="remaining 1 hour"
        >
          
          
          
          <small
            class="text-dark"
          >
            (0%)
          </small>
        </td>
         
        <td
          class="nowrap"
        >
          <span
            class="badge badge-state badge-success"
          >
            Running
          </span>
        </td>
         
        <td
          class="nowrap"
          title="started 2018-09-25 17:03"
        >
          2018-09-25 18:03
        </td>
         
        <td
          style="width: 600px;"
        >
          <div
            class="ellipsis ellipsis-indicator cursor"
          >
            <span
              class="text-muted nowrap mr-1"
            >
              (10 nodes)
            </span>
             
            <span
              class="nodes mr-1 comma"
            >
              a8-1.devgrenoble
            </span>
            <span
              class="nodes mr-1 comma"
            >
              a8-17.devgrenoble
            </span>
            <span
              class="nodes mr-1 comma"
            >
              a8-2.devgrenoble
            </span>
            <span
              class="nodes mr-1 comma"
            >
              a8-3.devgrenoble
            </span>
            <span
              class="nodes mr-1 comma"
            >
              a8-4.devgrenoble
            </span>
            <span
              class="nodes mr-1 comma"
            >
              a8-5.devgrenoble
            </span>
            <span
              class="nodes mr-1 comma"
            >
              a8-6.devgrenoble
            </span>
            <span
              class="nodes mr-1 comma"
            >
              a8-7.devgrenoble
            </span>
            <span
              class="nodes mr-1 comma"
            >
              a8-8.devgrenoble
            </span>
            <span
              class="nodes mr-1"
            >
              a8-9.devgrenoble
            </span>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
